### ISSUE TITLE
W^X non-compliance

### ISSUE THREAD

**Issue by mmcco** [racket/racket#1296]
I noticed in a [list maintained by a PaX developer](https://github.com/thestinger/paxd/blob/master/paxd.conf) that Racket doesn't yet conform to W^X. This means that Racket maps memory pages that at some point simultaneously have write and execute permissions at the hardware level. These permissions are determined by the options passed to `mmap(2)`, and by any subsequent calls to `mprotect(2)`.

WX pages make heap-based exploits easier, so they're being phased out. I'm not aware of a reasonable use case for them, either. Relatively few projects still use WX pages (most include as-of-yet untamed JIT compilers), and in the not-so-distant future operating systems will probably enforce W^X.

I don't have Racket development experience and don't have time to start at the moment. I thought I'd mention this to start the conversation, though.

A few examples:
- `racket/src/racket/src/salloc.c:941`
- `racket/src/racket/src/salloc.c:947`
- `racket/src/racket/sgc/sgc.c:975`
- `racket/src/racket/gc/os_dep.c:1940`

Thanks for your time,
Mike

**Comment by mflatt** [comment]
Thanks for raising this issue.

Do you know anything about know how other systems that mix libffi (for callbacks) and OS-level threads handle W^X? I wasn't able to find any hints by searching, so far.

**Comment by drbo** [comment]
The _general_ solution on Linux is to use a memory-mapping trick to map _two_ views of the same memory: one writable and one executable, such that writes to the writable mapping affect the executable mapping.

**Comment by mflatt** [comment]
@drbo Thanks! I should have added a link here to the discussion at https://groups.google.com/d/topic/racket-users/MtUMGmAacn0/discussion, which includes comments along those lines.

**Comment by mmcco** [comment]
@drbo Doesn't page aliasing defeat much of the mitigation? Full W^X prevents people from, for example, using out-of-bounds writes to directly change JIT-compiled code that is being executed. With page aliasing, attackers may be able to find the location of the aliased writeable page, at which point we're back where we started. The last chunk of this email has a good discussion of aliasing:

https://marc.info/?l=openbsd-misc&m=145944191229033&w=2

I'm pretty new to these topics, though, so I don't have much of value to offer.

**Comment by juanfra684** [comment]
Related: http://thread.gmane.org/gmane.os.openbsd.ports/79903

**Comment by pmatos** [comment]
@mflatt I have seen a few commits related to this: https://github.com/racket/racket/search?q=W%5EX&type=Commits

Is this already fixed?

**Comment by pmatos** [comment]
OK, this only seems to be implemented on OpenBSD x86_64.

**Comment by juanfra684** [comment]
@pmatos if you have an OS which enforces W^X, add `IMPLEMENT_WRITE_XOR_EXECUTE_BY_SIGNAL_HANDLER` to the correct block in `sconfig.h` and give it a try. There is nothing related to OpenBSD in the handler with the exception of the entry in `sconfig.h`.

**Comment by pmatos** [comment]
sigsegv in x86_64 linux. Does it actually work in OpenBSD or do we just assume it does?

**Comment by pmatos** [comment]
@juanfra684 you got me curious, is there an OpenBSD vagrant box available?

**Comment by mflatt** [comment]
It actually works on x86_64 OpenBSD. That may be the only combination where it works. I tried recently with x86_64 NetBSD, and it doesn't work there.

**Comment by pmatos** [comment]
Yes, I confirm it works on OpenBSD as well. Segfault on x86_64 linux - interesting!
https://gitlab.com/LinkiTools/racket/-/jobs/259795109

Might be interesting to point out that it's ok with SGC (builds and passes tests), but fails with 3m (builds but tests don't even run because it segfaults trying to install `racket-test`)

I have quite a bit on my plate now but I will get back to this another day - if nobody else does.

**Comment by juanfra684** [comment]
I don't know the technical details but the implementation of W^X used on NetBSD and HardenedBSD are based on the old PaX patch for the Linux kernel. IIRC, the OpenBSD implementation is slightly more relaxed to make it more compatible with third party software. For example, our Firefox package runs with W^X enforced but the NetBSD port requires disabling W^X to run.

OpenBSD would prefer to have a more strict W^X implementation but then you couldn't run any third party software. Be aware that in addition to W^X we have more protections than other OS, so that little tradeoff is not so critical on OpenBSD.

**Comment by sorawee** [comment]
@mflatt is this fixed by your recent change?

**Comment by mflatt** [comment]
No. See https://github.com/racket/racket/blob/master/racket/src/ChezScheme/c/segment.c#L571

**Comment by mflatt** [comment]
Some notes on the current status of W^X in Racket:

The thorniest problem remains the combination of OS-level threading and allocating/collecting code. When code can be allocated in one thread and run in another, then at least two things can go wrong:

 * New code may be allocated on a page where there's code already, so one thread may run code while another thread is trying to allocate code on the same page.  Although allocation is thread-local, that doesn't help if code allocated in one thread is communicated to another thread.

 * Foreign code in an unknown thread might invoke a callback into Racket/Scheme while the garbage collector is running. Although a lock will prevent the callback from running as the same time as the GC, the code that takes the lock is itself in a GCable space (but made immobile so that it isn't moved while it could be called).

One solution to these problem is for the OS to provide better support for W^X. An example is in Mac OS Big Sur for AArch64, where the W or X disposition of a page is thread-specific (and all pages that can change disposition have the same one for a given thread at any given time). Racket CS and BC implement W^X in that environment.

@Bogdanp made Racket CS work to a useful degree in a less cooperative environment (where his target was iOS). Page protection is flipped at a relatively fine granularity for non-GC updates, and a GC changes protections only for generations being collected. That reduces the chance of a thread conflict enough that it works in practice for his application.

As an example where it doesn't work well enough, a build on Racket of Mac OS tends to crash at the point where it uses the Mac SSL library to fetch packages (including the one that provides libopenssl), and using that library requires OS threads. If the build is moved past that hurdle, it tend to crash somewhat later as Cairo is used during parallel compilation of bytecode. Also, fine-grained permission flipping slows common Racket operations (such as expanding and compiling code) by 10-15% on Mac OS, although that doesn't necessarily say anything about how it performs on other OSes.

Overall, it doesn't look likely that Racket can support W^X without more OS support than being able to flip a page's permissions process-wide.

**Comment by samth** [comment]
I'm confused by the next-to-last paragraph -- does "a build of Racket on Mac OS" refer to something different than the prior paragraph about Big Sur?

**Comment by mflatt** [comment]
I mean that if I change CS to use Bogdan's implementation of W^X on Mac OS, then building Racket fails. (I tried on Big Sur, but I don't think the specific Mac OS version would matter.)

**Comment by Bogdanp** [comment]
Even on iOS, I ultimately hit a dead end. Everything works fine on iOS 14.3, but as of version 14.4, the application only works when it's attached to a debugger. Otherwise, it segfaults with a `KERN_CODESIGN_ERROR`. It seems like enforcement of dynamic code-signing requirements may have been broken for a time in iOS and I ended up inadvertently exploiting it. So the current state W^X on iOS is that it only works under XCode and on "jailbroken" devices.

**Comment by DemiMarie** [comment]
> One solution to these problem is for the OS to provide better support for W^X. An example is in Mac OS Big Sur for AArch64, where the W or X disposition of a page is thread-specific (and all pages that can change disposition have the same one for a given thread at any given time). Racket CS and BC implement W^X in that environment.

Another approach is to allocate executable memory entire pages at a time, and never move them afterwards.  That is wasteful of memory, though.

**Comment by DemiMarie** [comment]
> Even on iOS, I ultimately hit a dead end. Everything works fine on iOS 14.3, but as of version 14.4, the application only works when it's attached to a debugger. Otherwise, it segfaults with a [`KERN_CODESIGN_ERROR`](https://github.com/apple/darwin-xnu/blob/8f02f2a044b9bb1ad951987ef5bab20ec9486310/osfmk/vm/vm_fault.c#L2552-L2583). It seems like enforcement of dynamic code-signing requirements may have been broken for a time in iOS and I ended up inadvertently exploiting it. So the current state W^X on iOS is that it only works under XCode and on "jailbroken" devices.

iOS requires full ahead-of-time compilation.  This has other advantages too, like allowing the compiler to be stripped from the image.

**Comment by jbclements** [comment]
So maybe the release notes shouldn’t advertise that we have cross-compilation to iOS…

(edit edit edit).

John

**Comment by DemiMarie** [comment]
If we want to support iOS, we will need to avoid having to allocate executable memory at runtime.  We can still use libffi for callbacks, however.

**Comment by LiberalArtist** [comment]
For the very restricted case of iOS, would it be a viable approach to embed a Racket program as a Chez Scheme bootfile, or something similar?

As I understand it (not very thoroughly!), at the Chez Scheme level, the functionality in `petite.boot` wouldn't allocate code. Everything loaded from a boot file is in the static generation, which seems like it would avoid the GC issues. So it seems like, if a Racket program were able to be embedded as a boot file and to avoid depending on `scheme.boot` at runtime—or perhaps just to avoid using it—that would achieve ahead-of-time compilation for pure Racket code. The FFI is still a question: [CSUG](https://cisco.github.io/ChezScheme/csug9.5/use.html#g13) says that "compiled versions of `foreign-procedure` and `foreign-callable` forms may … be included in compiled code loaded into _Petite Chez Scheme_," but I think the [implementation](https://github.com/racket/racket/blob/0a9c70e95a69743dd5d219a395e995be4a4bfd41/racket/src/cs/rumble/foreign.ss#L1544-L1548) `eval`s them at runtime—maybe the memoization table could be prepopulated?

Even if this approach is viable, it seems like it would be a lot of work, and the result would have limited dynamic functionality. My preference, of course, would be for the platform to become more hospitable.

**Comment by DemiMarie** [comment]
> For the very restricted case of iOS, would it be a viable approach to embed a Racket program as a Chez Scheme bootfile, or something similar?

Perhaps.  I am not at all familiar with how that works, but I do know that anything that requires runtime `eval` will need to fall back to an interpreter.  Additionally, there may be other problems due to memory constraints on mobile devices.
